src/main/java/org/mockito/internal/verification/MockAwareVerificationMode.java

```diff
@@ -4,21 +4,44 @@
  */
 package org.mockito.internal.verification;
 
+import java.util.Set;
+
 import org.mockito.internal.verification.api.VerificationData;
+
+import org.mockito.listeners.VerificationListener;
+import org.mockito.verification.VerificationEvent;
 import org.mockito.verification.VerificationMode;
 
 public class MockAwareVerificationMode implements VerificationMode {
 
     private final Object mock;
     private final VerificationMode mode;
+    private final Set<VerificationListener> listeners;
 
-    public MockAwareVerificationMode(Object mock, VerificationMode mode) {
+    public MockAwareVerificationMode(Object mock, VerificationMode mode, Set<VerificationListener> listeners) {
         this.mock = mock;
         this.mode = mode;
+        this.listeners = listeners;
     }
 
     public void verify(VerificationData data) {
-        mode.verify(data);
+        try {
+            mode.verify(data);
+            notifyExceptionListeners(new VerificationEventImpl(mock, mode, data, null));
+        } catch (RuntimeException e) {
+            notifyExceptionListeners(new VerificationEventImpl(mock, mode, data, e));
+            throw e;
+        } catch (Error e) {
+            notifyExceptionListeners(new VerificationEventImpl(mock, mode, data, e));
+            throw e;
+        }
```

Comment:
Also I'd suggest two static factory that will create the concrete types : `VerificationSuccess` and `VerificationFailure`. That would look like : 

```
try {
    mode.verify(data);
    onVerification(VerificationFactory.success(mock, mode, data)); // build a VerificationSuccess class
} catch (... runtime) {
    onVerification(VerificationFactory.failure(mock, mode, data, runtime)); // create a VerificationFailure class
    throw runtime;
}
```